mapeditor/campaigneditor/campaigneditor.cpp

```diff
@@ -0,0 +1,252 @@
+	ui->actionCampaignProperties->setEnabled(campaignState != nullptr);
+
+	if(!campaignState)
+		return;
+
+	campaignScene->clear();
+
+	auto background = BitmapHandler::loadBitmap(campaignState->getRegions().getBackgroundName().getName());
+	campaignScene->addItem(new QGraphicsPixmapItem(QPixmap::fromImage(background)));
+	for (auto & s : campaignState->scenarios)
+	{
+		auto scenario = s.first;
+		auto color = campaignState->scenarios.at(scenario).regionColor;
+		auto image = BitmapHandler::loadBitmap(campaignState->getRegions().getAvailableName(scenario, color).getName());
+		if(selectedScenario == scenario)
+			image = BitmapHandler::loadBitmap(campaignState->getRegions().getSelectedName(scenario, color).getName());
+		else if(campaignState->scenarios.at(scenario).mapName == "")
+			image = BitmapHandler::loadBitmap(campaignState->getRegions().getConqueredName(scenario, color).getName());
+		auto pixmap = new ClickablePixmapItem(QPixmap::fromImage(image), [this, scenario]()
+		{
+			bool redrawRequired = selectedScenario != scenario;
+			selectedScenario = scenario;
+
+			if(redrawRequired)
+				redraw();
+		}, [this, scenario]()
+		{
+			if(ScenarioProperties::showScenarioProperties(campaignState, scenario))
+				changed();
+			redraw();
+		}, [this, scenario](QGraphicsSceneContextMenuEvent * event)
+		{
+			QMenu contextMenu(this);
+			QAction *actionScenarioProperties = contextMenu.addAction(tr("Scenario editor"));
+			actionScenarioProperties->setIcon(ui->actionScenarioProperties->icon());
+			connect(actionScenarioProperties, &QAction::triggered, this, [this, scenario]() {
+				if(ScenarioProperties::showScenarioProperties(campaignState, scenario))
+					changed();
+				redraw();
+			});
+			contextMenu.exec(event->screenPos());
+		});
+		auto pos = campaignState->getRegions().getPosition(scenario);
+		pixmap->setPos(pos.x, pos.y);
+		pixmap->setToolTip(QString::fromStdString(campaignState->scenarios.at(scenario).mapName));
+		campaignScene->addItem(pixmap);
+	}
+
+	ui->campaignView->show();
+}
+
+bool CampaignEditor::getAnswerAboutUnsavedChanges()
+{
+	if(unsaved)
+	{
+		auto sure = QMessageBox::question(this, tr("Confirmation"), tr("Unsaved changes will be lost, are you sure?"));
+		if(sure == QMessageBox::No)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+void CampaignEditor::setTitle()
+{
+	QFileInfo fileInfo(filename);
+	QString title = QString("%1%2 - %3 (%4)").arg(fileInfo.fileName(), unsaved ? "*" : "", VCMI_CAMP_EDITOR_NAME, GameConstants::VCMI_VERSION.c_str());
+	setWindowTitle(title);
+}
+
+void CampaignEditor::changed()
+{
+	unsaved = true;
+	setTitle();
+}
+
+bool CampaignEditor::saveCampaign()
+{
+	if(campaignState->mapPieces.size() != campaignState->campaignRegions.regions.size())
+	{
+		QMessageBox::critical(nullptr, tr("Maps missing"), tr("Not all Regions have a map. Please add them in Scenario Properties."));
+		return false;
+	}
```

Comment:
I think this is actually very legal case in H3 - for user-made campaigns you don't have to use all regions, some can be left empty